zkevm-circuits/src/evm_circuit.rs

```diff
@@ -1 +1,916 @@
+    GasAvailable,
+    // Value in wei of call.
+    Value,
+    // If call succeeds or not in the future.
+    IsPersistant,
+    // If call is within a static call.
+    IsStatic,
+    // If call is `CREATE*`. We lookup op from calldata when is create,
+    // otherwise we lookup code under address.
+    IsCreate,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum CallStateField {
+    // Program counter.
+    ProgramCounter,
+    // Stack pointer.
+    StackPointer,
+    // Memory size.
+    MemeorySize,
+    // Gas counter.
+    GasCounter,
+    // State trie write counter.
+    StateWriteCounter,
+    // Last callee's unique identifier.
+    CalleeID,
+    // Offset of returndata.
+    ReturndataOffset,
+    // Size of returndata.
+    ReturndataSize,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum BusMappingLookup<F> {
+    // Read-Write
+    OpExecutionState {
+        is_write: bool,
+        field: CallStateField,
+        value: Expression<F>,
+    },
+    Stack {
+        is_write: bool,
+        // One can only access its own stack, so id is no needed to be
+        // specified.
+        // Stack index is always deterministic respect to stack pointer, so an
+        // index offset is enough.
+        index_offset: i32,
+        value: Expression<F>,
+    },
+    Memory {
+        is_write: bool,
+        // Some ops like CALLDATALOAD can peek caller's memeory as calldata, so
+        // we allow it to specify the call id.
+        call_id: Expression<F>,
+        index: Expression<F>,
+        value: Expression<F>,
+    },
+    AccountStorage {
+        is_write: bool,
+        address: Expression<F>,
+        location: Expression<F>,
+        value: Expression<F>,
+        value_prev: Expression<F>,
+    },
+    // TODO: AccountNonce,
+    // TODO: AccountBalance,
+    // TODO: AccountCodeHash,
+
+    // Read-Only
+    Call {
+        field: CallField,
+        value: Expression<F>,
+    },
+    Bytecode {
+        hash: Expression<F>,
+        index: Expression<F>,
+        value: Expression<F>,
+    },
+    /* TODO: Block,
+     * TODO: Bytecode,
+     * TODO: Tx,
+     * TODO: TxCalldata, */
+}
+
+impl<F: FieldExt> BusMappingLookup<F> {
+    fn rw_target(&self) -> Target {
+        match self {
+            Self::Stack { .. } => Target::Stack,
+            Self::Memory { .. } => Target::Memory,
+            Self::AccountStorage { .. } => Target::Storage,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Eq, Ord)]
+pub(crate) enum FixedLookup {
+    // Noop provides [0, 0, 0, 0] row
+    Noop,
+    // meaningful tags start with 1
+    Range256,
+    Range32,
+    BitwiseAnd,
+    BitwiseOr,
+    BitwiseXor,
+}
+
+impl<F: FieldExt> Expr<F> for FixedLookup {
+    fn expr(&self) -> Expression<F> {
+        Expression::Constant(F::from_u64(*self as u64))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum Lookup<F> {
+    FixedLookup(FixedLookup, [Expression<F>; 3]),
```

Comment:
what does three expressions mean ? one is not enough ?

Reply:
The `FixedLookup` is the global 3-column fixed table with an extra tag column. It's used for range lookup, bitwise lookup, or anything linear-combination unfriendly.

Naively 3-columns table suits most common used table for 2 input and 1 output. We also put range table here to save columns, we might found the 3-columns table is not enough anymore, then we separate them out to another 1-column table to not waste cells.